internal/servers/repository_workerauth.go

```diff
@@ -0,0 +1,754 @@
+func (r *WorkerAuthRepositoryStorage) Store(ctx context.Context, msg nodee.MessageWithId) error {
+	const op = "servers.(WorkerAuthRepositoryStorage).Store"
+	if err := types.ValidateMessage(msg); err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	if msg.GetId() == "" {
+		return errors.New(ctx, errors.InvalidParameter, op, "given message cannot be stored as it has no ID")
+	}
+
+	// Determine type of message to store
+	switch t := msg.(type) {
+	case *types.NodeInformation:
+		err := r.storeNodeInformation(ctx, t)
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+	case *types.RootCertificates:
+		err := r.storeRootCertificates(ctx, t)
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+	default:
+		return errors.New(ctx, errors.InvalidParameter, op, fmt.Sprintf("message type %T not supported for Store", msg))
+	}
+
+	return nil
+}
+
+// Node information is stored in two parts:
+// * the workerAuth record is stored with a reference to a worker
+// * certificate bundles are stored with a reference to the workerAuth record and issuing root certificate
+func (r *WorkerAuthRepositoryStorage) storeNodeInformation(ctx context.Context, node *types.NodeInformation) error {
+	const op = "servers.(WorkerAuthRepositoryStorage).storeNodeInformation"
+	if node == nil {
+		return errors.New(ctx, errors.InvalidParameter, op, "missing NodeInformation")
+	}
+
+	nodeAuth := allocWorkerAuth()
+	nodeAuth.WorkerKeyIdentifier = node.Id
+	nodeAuth.WorkerEncryptionPubKey = node.EncryptionPublicKeyBytes
+	nodeAuth.WorkerSigningPubKey = node.CertificatePublicKeyPkix
+
+	// Encrypt the private key
+	databaseWrapper, err := r.kms.GetWrapper(ctx, scope.Global.String(), kms.KeyPurposeDatabase)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	nodeAuth.KeyId, err = databaseWrapper.KeyId(ctx)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	nodeAuth.ControllerEncryptionPrivKey, err = encrypt(ctx, node.ServerEncryptionPrivateKeyBytes, databaseWrapper)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	nodeAuth.Nonce, err = encrypt(ctx, node.RegistrationNonce, databaseWrapper)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+
+	// Get workerId from state passed in
+	var result workerAuthWorkerId
+	err = mapstructure.Decode(node.State.AsMap(), &result)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	nodeAuth.WorkerId = result.WorkerId
+
+	if err := nodeAuth.ValidateNewWorkerAuth(ctx); err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+
+	_, err = r.writer.DoTx(ctx, db.StdRetryCnt, db.ExpBackoff{}, func(read db.Reader, w db.Writer) error {
+		// Store WorkerAuth
+		storeNodeAuth := nodeAuth.clone()
+		if err := w.Create(ctx, &storeNodeAuth); err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		// Then store cert bundles associated with this WorkerAuth
+		for _, c := range node.CertificateBundles {
+			if err := r.storeWorkerCertBundle(ctx, c, nodeAuth.WorkerKeyIdentifier, w); err != nil {
```

Comment:
This is an empty branch -- it will do nothing and return nil rather than abort the transaction